### ISSUE TITLE
Infer key-of constant array based on array_key_exists check

### ISSUE THREAD

**Issue by bdsl** [vimeo/psalm#1830]
I would like the following code to be valid. Currently Psalm emits `ArgumentTypeCoercion - 28:31 - Argument 1 of foo::dogetdayname expects int(1)|int(2)|int(3)|int(4)|int(5)|int(6)|int(7), parent type int provided`

```
<?php
  
class foo
{
  /** @var array<int,string> */
  public const DAYS = [
    1 => 'mon',
    2 => 'tue',
    3 => 'wed',
    4 => 'thu',
    5 => 'fri',
    6 => 'sat',
    7 => 'sun'
  ];
  
  /** @param key-of<self::DAYS> $dayNum*/
  private static function doGetDayName(int $dayNum): string
  {
    return self::DAYS[$dayNum];
  }
  
  /** @throws LogicException */
  public static function getDayName(int $dayNum): string
  {
    if (! array_key_exists($dayNum, self::DAYS)) {
      throw new \LogicException();
    }
    return self::doGetDayName($dayNum); // ArgumentTypeCoercion emitted here
  }
}
```
https://psalm.dev/r/8afe8ba4ea

For now we can work around this by adding `/** @var key-of<self::DAYS> $dayNum */` (https://psalm.dev/r/7cc1f58d92 )

**Comment by weirdan** [comment]
Related: #942

**Comment by bdsl** [comment]
Fixed in release [3.6.5](https://github.com/vimeo/psalm/releases/tag/3.6.5)
